File: backend/opencode/providers.go
package opencode

const (
	ProviderOpenCode = "opencode-go"
	ProviderCodex    = "openai-codex"
)

// ProviderInfo describes a model provider shown in the /model picker.
type ProviderInfo struct {
	ID   string
	Name string
}

// ModelProviders returns the providers available in the model picker.
func ModelProviders() []ProviderInfo {
	return []ProviderInfo{
		{ID: ProviderOpenCode, Name: "OpenCode Go"},
		{ID: ProviderCodex, Name: "OpenAI Codex"},
	}
}

// codexModelOrder mirrors Hermes' curated Codex OAuth catalog (chatgpt.com backend).
var codexModelOrder = []string{
	"gpt-5.5",
	"gpt-5.4-mini",
	"gpt-5.4",
	"gpt-5.3-codex",
	"gpt-5.3-codex-spark",
	"gpt-5-codex",
}

var codexKnownModels = map[string]ModelInfo{
	"gpt-5.5": {
		ID: "gpt-5.5", Name: "GPT-5.5",
		ContextWindow: 272_000, Reasoning: true,
	},
	"gpt-5.4-mini": {
		ID: "gpt-5.4-mini", Name: "GPT-5.4 Mini",
		ContextWindow: 272_000, Reasoning: true,
	},
	"gpt-5.4": {
		ID: "gpt-5.4", Name: "GPT-5.4",
		ContextWindow: 272_000, Reasoning: true,
	},
	"gpt-5.3-codex": {
		ID: "gpt-5.3-codex", Name: "GPT-5.3 Codex",
		ContextWindow: 272_000, Reasoning: true,
	},
	"gpt-5.3-codex-spark": {
		ID: "gpt-5.3-codex-spark", Name: "GPT-5.3 Codex Spark",
		ContextWindow: 128_000, Reasoning: true,
	},
	"gpt-5-codex": {
		ID: "gpt-5-codex", Name: "GPT-5 Codex",
		ContextWindow: 272_000, Reasoning: true,
	},
}

// CodexModels returns the static Codex OAuth model catalog.
func CodexModels() []ModelInfo {
	out := make([]ModelInfo, 0, len(codexModelOrder))
	for _, id := range codexModelOrder {
		var m ModelInfo
		if known, ok := codexKnownModels[id]; ok {
			m = known
		} else {
			m = ModelInfo{ID: id, Name: id, Reasoning: true, ContextWindow: 272_000}
		}
		m.ThinkingLevels = append([]ThinkingLevel(nil), defaultReasoningLevels...)
		out = append(out, normalizeModel(m))
	}
	return out
}

// ModelsForProvider returns models for a provider, using the registry for OpenCode.
func ModelsForProvider(provider string, registry *Registry) []ModelInfo {
	switch provider {
	case ProviderCodex:
		if registry != nil {
			return registry.CodexModelsList()
		}
		return CodexModels()
	default:
		models := registry.Models()
		if len(models) == 0 {
			models = FallbackModels()
		}
		out := make([]ModelInfo, len(models))
		copy(out, models)
		sortModels(out)
		return out
	}
}

// LookupCatalogModel resolves model metadata from OpenCode or Codex catalogs.
func LookupCatalogModel(id string) (ModelInfo, bool) {
	if m, ok := LookupModel(id); ok {
		return m, true
	}
	if m, ok := defaultRegistry.LookupCodex(id); ok {
		return m, true
	}
	return ModelInfo{}, false
}

// ProviderIndex returns the index of provider id in ModelProviders, or 0.
func ProviderIndex(provider string) int {
	for i, p := range ModelProviders() {
		if p.ID == provider {
			return i
		}
	}
	return 0
}

File: backend/opencode/providers.ts
// PORT: backend/opencode/providers.go

export const provider_opencode = "opencode-go";
export const provider_opencode_zen = "opencode-zen";
export const provider_neuralwatt = "neuralwatt";
export const provider_codex = "openai-codex";

export type provider_info = { id: string; name: string };
export type thinking_level = { id: string; name: string };
export type model_info = {
  id: string;
  name: string;
  context_window: number;
  reasoning: boolean;
  supports_images?: boolean;
  thinking_levels?: thinking_level[];
  reasoning_field?: string;
  mandatory_thinking?: boolean;
};

export const default_reasoning_levels: thinking_level[] = [];

export const model_providers = (): provider_info[] => [
  { id: provider_opencode, name: "OpenCode Go" },
  { id: provider_opencode_zen, name: "OpenCode Zen" },
  { id: provider_neuralwatt, name: "NeuralWatt" },
  { id: provider_codex, name: "OpenAI Codex" },
];

import {
  resolve_model,
  codex_fallback_ids,
} from "./model_resolver";
import {
  sort_models,
  fallback_models,
} from "./models";

export const codex_models = (): model_info[] => {
  return codex_fallback_ids.map((id) => resolve_model(id, provider_codex));
};

export type registry_like = {
  codex_models_list?: () => model_info[];
  zen_models_list?: () => model_info[];
  neuralwatt_models_list?: () => model_info[];
  models?: () => model_info[];
  lookup_neuralwatt?: (id: string) => [model_info, boolean];
  lookup_codex?: (id: string) => [model_info, boolean];
};

export const models_for_provider = (provider: string, registry: registry_like | null): model_info[] => {
  switch (provider) {
    case provider_codex:
      return registry?.codex_models_list?.() ?? codex_models();
    case provider_opencode_zen: {
      const out = registry?.zen_models_list?.() ?? fallback_models(provider_opencode_zen);
      sort_models(out); return out;
    }
    case provider_neuralwatt: {
      const out = registry?.neuralwatt_models_list?.() ?? [];
      sort_models(out); return out;
    }
    default: {
      let out = registry?.models?.() ?? fallback_models(provider_opencode);
      if (out.length === 0) {
        out = fallback_models(provider_opencode);
      }
      sort_models(out);
      return out;
    }
  }
};

export const lookup_catalog_model = (id: string): [model_info, boolean] => {
  const clean = id.trim();
  if (clean === "") {
    return [{ id: "", name: "", context_window: 0, reasoning: false }, false];
  }
  const m = resolve_model(clean, provider_opencode);
  if (m.id !== "") {
    return [m, true];
  }
  return [{ id: "", name: "", context_window: 0, reasoning: false }, false];
};

export const provider_index = (provider: string): number => {
  const providers = model_providers();
  const idx = providers.findIndex((p) => p.id === provider);
  return idx < 0 ? 0 : idx;
};

/*
PORT STATUS
source path: backend/opencode/providers.go
source lines: 145
draft lines: 111
confidence: high
status: phase_b_compile
*/
